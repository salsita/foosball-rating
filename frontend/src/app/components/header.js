import React, { Component } from 'react';
import {
  Title, Subtitle,
  Logo,
  Nav
} from './../../styles/blocks/';

const logo = require('./../../media/logo.png');

class Header extends Component {
  render() {
    return(
      <Nav>
        <Logo href="https://www.salsitasoft.com/"><img src={logo} /></Logo>
        <Title>ELO rating</Title>
      </Nav>
    )
  }
}

export default Header;
