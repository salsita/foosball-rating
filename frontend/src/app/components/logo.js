import React from 'react'
import logoSmall from './../../media/logo-small.png'
import logo from './../../media/logo.png'
import { Link } from 'react-router-dom'

export const LogoComponent = ({ className }) => (
  <span className={className}>
    <Link id='smalllogo' to={'/'}>
      <img src={logoSmall} alt="logo" />
    </Link>
    <Link id='largelogo' to={'/'}>
      <img src={logo} alt="logo" />
    </Link>
  </span>
)
