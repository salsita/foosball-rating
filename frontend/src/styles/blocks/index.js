import { Title, Subtitle, TextSpan, WinnerSpan, BattleLabel, StyledHyperLink } from './typo'
import { Nav, Container, GridContainer, Box, ProfileDetail } from './layout'
import { ListCon, ListItem } from './list'
import { StyledLink } from './link'
import Button from './buttons'
import { SelectBox, Label, Input } from './inputs'
import Logo from './logo'

export {
  Title, Subtitle, TextSpan, WinnerSpan, BattleLabel, StyledHyperLink,
  Nav, Container, GridContainer, Box, ProfileDetail,
  ListCon, ListItem,
  StyledLink,
  Button,
  SelectBox, Label, Input,
  Logo,
}
