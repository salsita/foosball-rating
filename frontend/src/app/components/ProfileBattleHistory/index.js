import React from 'react'
import { ListCon } from './../../../styles/blocks'
import { ProfileBattleHistoryRow } from './ProfileBattleHistoryRow'
import { TextSpan } from '../../../styles/blocks/typo'

const ProfileBattleHistory = props => (
  <>
    <TextSpan>
      Match history
    </TextSpan>
    <ListCon>
      {props.matches.map(match =>
        <ProfileBattleHistoryRow key={match.id} match={match} />,
      )}
    </ListCon>
  </>
)

export default ProfileBattleHistory
