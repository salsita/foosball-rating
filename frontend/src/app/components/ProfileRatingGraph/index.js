import React from 'react'
import { connect } from 'react-redux'
import {
  FlexibleWidthXYPlot,
  XAxis,
  YAxis,
  HorizontalGridLines,
  VerticalGridLines,
  LineSeries,
} from 'react-vis'

import { getRatingHistoryGraphForPlayer } from '../../modules/matches/matches-selectors'
import { Box, TextSpan } from '../../../styles/blocks'
import { plotAxisStyle, plotGridStyle, plotLineStyle, plotMainGridStyle } from '../../../styles/svg'

const SimpleProfileRatingGraph = props => (
  <Box Margin='0 5px 20px 5px'>
    <TextSpan>
      Rating history
    </TextSpan>
    <div>
      <FlexibleWidthXYPlot
        height={300}
        xType='time'
        margin={{ left: 45 }}
      >
        <HorizontalGridLines
          style={plotGridStyle}
        />
        <HorizontalGridLines
          style={plotMainGridStyle}
          tickValues={[1000]}
        />
        <VerticalGridLines />
        <XAxis
          style={plotAxisStyle}
          tickLabelAngle={-45}
        />
        <YAxis style={plotAxisStyle} />
        <LineSeries
          data={props.ratingHistoryGraph}
          style={plotLineStyle}
        />
      </FlexibleWidthXYPlot>
    </div>
  </Box>
)


const mapStateToProps = (state, props) => ({
  ratingHistoryGraph: getRatingHistoryGraphForPlayer(state, props.playerId),
})

export const ProfileRatingGraph = connect(mapStateToProps)(SimpleProfileRatingGraph)
