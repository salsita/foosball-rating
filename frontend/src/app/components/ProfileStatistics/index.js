import React from 'react'

import { TextSpan, StatisticsTable, StatisticsValue, StyledLink, WinnerSpan } from '../../../styles/blocks'
import { createProfilePath } from '../../const/routes'

const ProfileStatistics = props => {
  const {
    rankings: { ranking, toNextRank, toPrevRank },
    statistics: { 
      bestDay, worstDay, mostFrequentTeammate, leastFrequentTeammate, mostFrequentOpponent, 
      leastFrequentOpponent, mostSuccessTeammate, leastSuccessTeammate, mostSuccessOpponent, 
      leastSuccessOpponent,
    },
  } = props

  return (
    <>
      <TextSpan>Ranking: {ranking}</TextSpan>
      <StatisticsTable>
        <tbody>
          <tr>
            <td>To Next Rank: 
              <StatisticsValue>{getScorePart(toNextRank, toNextRank <= 20, '+')}</StatisticsValue>
            </td>
            <td>To Prev Rank: 
              <StatisticsValue>{getScorePart(toPrevRank, toPrevRank > 20, '-')}</StatisticsValue>
            </td>
          </tr>
          <tr>
            <td>Best Day: 
              <StatisticsValue>{bestDay?.date} {getDayScorePart(bestDay?.value)}</StatisticsValue>
            </td>
            <td>Worst Day: 
              <StatisticsValue>{worstDay?.date} {getDayScorePart(worstDay?.value)}</StatisticsValue>
            </td>
          </tr>
          <tr>
            {getUserRow('Most Played With', mostFrequentTeammate, 'matches', true)}
            {getUserRow('Least PLayed With', leastFrequentTeammate, 'matches', false)}
          </tr>
          <tr>
            {getUserRow('Most Played Against', mostFrequentOpponent, 'matches', true)}
            {getUserRow('Least Played Against', leastFrequentOpponent, 'matches', false)}
          </tr>
          <tr>
            {getUserRow('Most Won With', mostSuccessTeammate, 'wins', true)}
            {getUserRow('Least Won With', leastSuccessTeammate, 'wins', false)}
          </tr>
          <tr>
            {getUserRow('Most Won Over', mostSuccessOpponent, 'wins', true)}
            {getUserRow('Most Lost To', leastSuccessOpponent, 'losses', false)}
          </tr>
        </tbody>
      </StatisticsTable>
    </>
  )
}

const getDayScorePart = score => getScorePart(score, score > 0, score > 0 ? '+' : '')

const getScorePart = (value, isGreen, sign = 'x') => (
  <WinnerSpan winner={isGreen}>{sign}{value}</WinnerSpan>
)

const getUserRow = (text, stat, valueKey, isGreen) => (
<td>{text}: 
  <StatisticsValue>
    {getPlayerLink(stat.user)} {getScorePart(stat[valueKey], isGreen)}
  </StatisticsValue>
</td>
)

const getPlayerLink = player => (
  <StyledLink key={player?.id} to={createProfilePath(player?.id)}>{player?.name} 
   &nbsp;<span>({player?.matchRating})</span></StyledLink>
)

export default ProfileStatistics
