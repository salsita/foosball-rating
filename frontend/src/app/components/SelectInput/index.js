import React from 'react'
import { SelectBox } from './../../../styles/blocks'

const SelectInput = ({ items, selectedValue, onChange }) => (
  <SelectBox onChange={onChange} value={selectedValue}>
    {items.map(item =>
      <option key={item.value} value={item.value}>
        {item.label}
      </option>,
    )}
  </SelectBox>
)

export default SelectInput


