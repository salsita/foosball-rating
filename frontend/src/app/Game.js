import React, { Component } from 'react'
import { Route, Switch } from 'react-router-dom'

import * as ROUTES from './const/routes'
import { Header } from './components/header'
import { Footer } from './components/Footer/footer'
// pages
import { Dashboard } from './pages/Dashboard'
import { LeaderboardPage } from './pages/Leaderboard'
import { CreateMatchPage } from './pages/CreateMatch'
import { Profile } from './pages/Profile'
import { AddPlayerPage } from './pages/AddPlayer'
import { MatchListPage } from './pages/MatchList'
import { SimpleButton, Subtitle, SimpleLink, Container } from '../styles/blocks'
import { connect } from 'react-redux'
import urljoin from 'url-join'
import { SelectionStatus } from './const/games'
import { capitalize } from './utils/string-utils'

export class GameComponent extends Component {
  render() {
    const { match: { url }, selection } = this.props
    const constructUrl = relativePath => urljoin(url, relativePath)
    const createMatch = () => {
      this.props.history.push(constructUrl(ROUTES.CREATE_MATCH))
    }
    return (
      <>
        <Header>
          { selection.status === SelectionStatus.SELECTED
            ? <>
              <span id='title'>
                <SimpleLink to={url}>{capitalize(selection.value.name)}</SimpleLink>
              </span>
              <SimpleButton onClick={createMatch}>Add Match</SimpleButton>
            </>
            : null }
        </Header>
        <Container id="container">
          { selection.status === SelectionStatus.SELECTED
            ? <Switch>
              <Route exact path={constructUrl(ROUTES.LEADERBOARD)}>
                <LeaderboardPage constructUrl={constructUrl} />
              </Route>
              <Route exact path={constructUrl(ROUTES.CREATE_MATCH)}>
                <CreateMatchPage constructUrl={constructUrl} />
              </Route>
              <Route exact path={constructUrl(ROUTES.ADD_PLAYER)} component={AddPlayerPage} />
              <Route exact path={constructUrl(ROUTES.MATCH_LIST)}>
                <MatchListPage constructUrl={constructUrl} />
              </Route>
              <Route exact path={constructUrl(ROUTES.PROFILE)} component={Profile} />
              <Route>
                <Dashboard constructUrl={constructUrl} />
              </Route>
            </Switch>
            : <Subtitle>
              { selection.status === SelectionStatus.FAILED
                ? `Game '${selection.gameName}' was not found!`
                : 'Loading...' }
            </Subtitle>
          }
        </Container>
        <Footer />
      </>
    )
  }
}

const mapStateToProps = state => ({
  selection: state.gameSelection,
})

export const Game = connect(mapStateToProps)(GameComponent)
