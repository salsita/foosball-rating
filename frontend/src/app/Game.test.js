import { GameComponent } from './Game'
import { Header } from './components/header'
import { Button } from '../styles/blocks'
import { shallowWithProps } from '../tests/utils'
import { HISTORY_MOCK, ROUTER_MATCH } from '../tests/mocks'
import { FOOSBALL_GAME, URLS } from '../tests/data'
import {
  createFailedSelection,
  createGameSelection,
  createEmptySelection,
} from './types/GameSelection'

describe('GameComponent', () => {
  let gameComponent
  beforeEach(() => {
    jest.clearAllMocks()
  })
  describe('when foosball selection fails', () => {
    beforeEach(() => {
      gameComponent = shallowWithProps(GameComponent, {
        match: { },
        selection: createFailedSelection(FOOSBALL_GAME.name),
      })
    })
    it('renders with no button and message "Game \'foosball\'" was not found', () => {
      expect(gameComponent).toMatchSnapshot()
    })
  })

  describe('when foosball is selected', () => {
    beforeEach(() => {
      gameComponent = shallowWithProps(GameComponent, {
        match: { ...ROUTER_MATCH, url: URLS.FOOSBALL },
        selection: createGameSelection(FOOSBALL_GAME),
        history: HISTORY_MOCK,
      })
    })
    it('renders button Add Match in the header', () => {
      expect(gameComponent).toMatchSnapshot()
    })
    describe('on Add Match button click', () => {
      beforeEach(() => {
        gameComponent.find(Header).find(Button).simulate('click')
      })
      it('redirects to foosball create match page', () => {
        expect(HISTORY_MOCK.push).toBeCalledWith(URLS.CREATE_FOOSBALL_MATCH)
      })
    })
  })

  describe('when the selection is empty', () => {
    beforeEach(() => {
      gameComponent = shallowWithProps(GameComponent, {
        match: { ...ROUTER_MATCH, url: URLS.FOOSBALL },
        selection: createEmptySelection(),
      })
    })
    it('renders header with no button and with message "Loading"', () => {
      expect(gameComponent).toMatchSnapshot()
    })
  })
})
