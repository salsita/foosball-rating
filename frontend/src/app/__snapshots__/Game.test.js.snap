// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameComponent when foosball selection fails renders with no button and message "Game 'foosball'" was not found 1`] = `
<Fragment>
  <Connect(HeaderComponent) />
  <styled.div>
    <styled.h2>
      Game 'foosball' was not found!
    </styled.h2>
  </styled.div>
  <Footer />
</Fragment>
`;

exports[`GameComponent when the selection is empty renders header with no button and with message "Loading" 1`] = `
<Fragment>
  <Connect(HeaderComponent) />
  <styled.div>
    <styled.h2>
      Loading...
    </styled.h2>
  </styled.div>
  <Footer />
</Fragment>
`;

exports[`GameComponent when foosball is selected renders button Add Match in the header 1`] = `
<Fragment>
  <Connect(HeaderComponent)>
    <styled.button
      onClick={[Function]}
    >
      Add Match
    </styled.button>
  </Connect(HeaderComponent)>
  <styled.div>
    <Switch>
      <Route
        exact={true}
        path="/foosball/leaderboard"
      >
        <Connect(LeaderboardPageComponent)
          constructUrl={[Function]}
        />
      </Route>
      <Route
        exact={true}
        path="/foosball/create-match"
      >
        <CreateMatchPage
          constructUrl={[Function]}
        />
      </Route>
      <Route
        component={[Function]}
        exact={true}
        path="/foosball/add-player"
      />
      <Route
        exact={true}
        path="/foosball/match-list"
      >
        <Connect(MatchListPageComponent)
          constructUrl={[Function]}
        />
      </Route>
      <Route
        component={[Function]}
        exact={true}
        path="/foosball/profile/:playerId"
      />
      <Route>
        <Connect(withRouter(DashboardComponent))
          constructUrl={[Function]}
        />
      </Route>
    </Switch>
  </styled.div>
  <Footer />
</Fragment>
`;
